File: ios/apis/JarvisApi.swift
import Foundation

class JarvisApi {
    
    var apiHost = "https://jarvis.viatick.com/apis"
    
    func getApplicationDetail(sdkKey: String)-> Void {
        var url = URL(string: apiHost + "/account/application/detail")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("Access-Token", forHTTPHeaderField: sdkKey)
        
        let applicationDetail = URLSession.shared.dataTask(with: request) {data, response, error in
            guard
                let data = data,
                let response = response as? HTTPURLResponse,
                error == nil
            else {                                                               // check for fundamental networking error
                print("error", error ?? URLError(.badServerResponse))
                return
            }
            
            guard (200 ... 299) ~= response.statusCode else {                    // check for http errors
                print("statusCode should be 2xx, but is \(response.statusCode)")
                print("response = \(response)")
                return
            }
            
            // do whatever you want with the `data`, e.g.:
            
            do {
                let responseObject = try JSONDecoder().decode(ApplicationDetail.self, from: data)
                print(responseObject)
            } catch {
                print(error) // parsing error
                
                if let responseString = String(data: data, encoding: .utf8) {
                    print("responseString = \(responseString)")
                } else {
                    print("unable to parse response as string")
                }
            }
        }
        applicationDetail.resume()
    }
    
    func findNotificationByDevice(sdkKey: String, uuid: String, major: Int, minor: Int)-> Void {
        var url = URL(string: apiHost + "/resource/locating-notification/find-by-device")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("Access-Token", forHTTPHeaderField: sdkKey)
        let jsonData: [String: Any] = ["uuid": uuid, "major": major, "minor": minor]
        let body = try?JSONSerialization.data(withJSONObject: jsonData)
        request.httpBody = body
        
        let notification = URLSession.shared.dataTask(with: request) {data, response, error in
            guard
                let data = data,
                let response = response as? HTTPURLResponse,
                error == nil
            else {                                                               // check for fundamental networking error
                print("error", error ?? URLError(.badServerResponse))
                return
            }
            
            guard (200 ... 299) ~= response.statusCode else {                    // check for http errors
                print("statusCode should be 2xx, but is \(response.statusCode)")
                print("response = \(response)")
                return
            }
            
            // do whatever you want with the `data`, e.g.:
            
            do {
                let responseObject = try JSONDecoder().decode(ApplicationDetail.self, from: data)
                print(responseObject)
            } catch {
                print(error) // parsing error
                
                if let responseString = String(data: data, encoding: .utf8) {
                    print("responseString = \(responseString)")
                } else {
                    print("unable to parse response as string")
                }
            }
        }
        notification.resume()
    }
}

File: android/src/main/java/com/reactnativejarvistemplateappsdk/apis/JarvisApi.java
package com.reactnativejarvistemplateappsdk.apis;

import android.util.Log;

import com.reactnativejarvistemplateappsdk.apis.response.ApplicationDetail;
import com.reactnativejarvistemplateappsdk.apis.response.DeviceFilter;
import com.reactnativejarvistemplateappsdk.apis.response.LocatingNotification;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public class JarvisApi {

  private final String apiHost = "https://jarvis.viatick.com/apis";
  private final OkHttpClient httpClient = new OkHttpClient();

  private static final MediaType JSON_MEDIA_TYPE = MediaType.parse("application/json; charset=utf-8");

  private JarvisApi() {
  }

  public ApplicationDetail getApplicationDetail(String sdkKey) {
    Request request = new Request.Builder()
        .header("Access-Token", sdkKey)
        .url(apiHost + "/account/application/detail")
        .build();

    try {
      Response rp = this.httpClient.newCall(request).execute();

      if (rp.isSuccessful()) {
        String bodyResponse = rp.body().string();
        JSONObject responseObject = new JSONObject(bodyResponse);

        long id = responseObject.getLong("id");
        String name = responseObject.getString("name");
        String email = responseObject.getString("email");
        JSONObject deviceFilterJson = responseObject.getJSONObject("deviceFilter");
        JSONObject beaconFilterJson = deviceFilterJson.getJSONObject("attendance_beacon");
        String uuid = beaconFilterJson.getString("uuid");
        int major = beaconFilterJson.getInt("major");

        Map<String, DeviceFilter> deviceFilterMap = new HashMap<>();
        deviceFilterMap.put("attendance_beacon", new DeviceFilter(uuid, major));

        return new ApplicationDetail(id, name, email, deviceFilterMap);
      } else {
        Log.d("getApplicationDetail", "code" + rp.code());
      }
    } catch (IOException | JSONException e) {
      Log.e("JarvisApi", e.getLocalizedMessage());
    }

    return null;
  }

  public LocatingNotification findNotificationByDevice(String sdkKey, String uuid, int major, int minor) {
    JSONObject bodyJson = new JSONObject();
    try {
      bodyJson.put("uuid", uuid);
      bodyJson.put("major", major);
      bodyJson.put("minor", minor);
    } catch (JSONException e) {
      Log.e("JarvisApi", e.getLocalizedMessage());
    }

    RequestBody body = RequestBody.create(bodyJson.toString(), JSON_MEDIA_TYPE);
    Request request = new Request.Builder()
        .header("Access-Token", sdkKey)
        .url(apiHost + "/resource/locating-notification/find-by-device")
        .post(body)
        .build();

    try {
      Response rp = this.httpClient.newCall(request).execute();

      if (rp.isSuccessful()) {
        String bodyResponse = rp.body().string();
        JSONObject responseObject = new JSONObject(bodyResponse);

        long notificationId = responseObject.getLong("id");
        String title = responseObject.getString("title");
        String description = responseObject.getString("description");

        return new LocatingNotification(notificationId, title, description);
      } else {
        Log.d("findByDevice", "code" + rp.code());
      }
    } catch (IOException | JSONException e) {
      Log.e("JarvisApi", e.getLocalizedMessage());
    }

    return null;
  }

  public static JarvisApi getInstance() {
    return SingletonHelper.INSTANCE;
  }

  private static class SingletonHelper {
    private static final JarvisApi INSTANCE = new JarvisApi();
  }

}
